File: forms/genetics-assessment/full-stack-with-loco-tera-htmx-alpine/src/engine/genetics_grader.rs
use super::flagged_issues::detect_additional_flags;
use super::genetics_rules::all_rules;
use super::types::{
    AssessmentData, FiredRule, GraderContext, GradingResult, ManchesterInputs, Relative,
};
use super::utils::max_risk_level;

// Manchester scoring weights (Evans 2004 / Evans & Lalloo).
const MAN_W_FEMALE_BREAST_UNDER30: i32 = 6;
const MAN_W_FEMALE_BREAST_30_39: i32 = 4;
const MAN_W_FEMALE_BREAST_40_49: i32 = 3;
const MAN_W_MALE_BREAST: i32 = 8;
const MAN_W_OVARIAN_UNDER60: i32 = 8;
const MAN_W_PANCREATIC_UNDER60: i32 = 1;
const MAN_W_PROSTATE_UNDER60: i32 = 1;

fn nzi(n: Option<i32>) -> i32 {
    n.unwrap_or(0)
}

/// Compute the Manchester score for BRCA1/2 from per-cancer counts.
pub fn calculate_manchester_score(d: &AssessmentData) -> i32 {
    let m: &ManchesterInputs = &d.targeted_risk_scoring.manchester;
    let mut s = 0;
    // Proband contributions
    s += nzi(m.proband_female_breast_under30) * MAN_W_FEMALE_BREAST_UNDER30;
    s += nzi(m.proband_female_breast30to39) * MAN_W_FEMALE_BREAST_30_39;
    s += nzi(m.proband_female_breast40to49) * MAN_W_FEMALE_BREAST_40_49;
    s += nzi(m.proband_ovarian_under60) * MAN_W_OVARIAN_UNDER60;
    s += nzi(m.proband_male_breast) * MAN_W_MALE_BREAST;
    // Relative contributions
    s += nzi(m.relative_female_breast_under30) * MAN_W_FEMALE_BREAST_UNDER30;
    s += nzi(m.relative_female_breast30to39) * MAN_W_FEMALE_BREAST_30_39;
    s += nzi(m.relative_female_breast40to49) * MAN_W_FEMALE_BREAST_40_49;
    s += nzi(m.relative_ovarian_under60) * MAN_W_OVARIAN_UNDER60;
    s += nzi(m.relative_male_breast) * MAN_W_MALE_BREAST;
    s += nzi(m.relative_pancreatic_under60) * MAN_W_PANCREATIC_UNDER60;
    s += nzi(m.relative_prostate_under60) * MAN_W_PROSTATE_UNDER60;
    s
}

/// Count how many of the five Revised Bethesda criteria are met.
pub fn count_bethesda_met(d: &AssessmentData) -> i32 {
    let b = &d.targeted_risk_scoring.bethesda;
    let mut n = 0;
    if b.crc_under50 == "yes" {
        n += 1;
    }
    if b.synchronous_metachronous == "yes" {
        n += 1;
    }
    if b.msi_histology == "yes" {
        n += 1;
    }
    if b.first_degree_lynch_tumour == "yes" {
        n += 1;
    }
    if b.multiple_relatives_lynch == "yes" {
        n += 1;
    }
    n
}

/// Flatten every relative in the pedigree.
pub fn flatten_pedigree(d: &AssessmentData) -> Vec<&Relative> {
    let fp = &d.family_pedigree;
    let mut out: Vec<&Relative> = vec![
        &fp.maternal_grandmother,
        &fp.maternal_grandfather,
        &fp.paternal_grandmother,
        &fp.paternal_grandfather,
        &fp.mother,
        &fp.father,
    ];
    out.extend(fp.maternal_aunts_uncles.iter());
    out.extend(fp.paternal_aunts_uncles.iter());
    out.extend(fp.siblings.iter());
    out.extend(fp.children.iter());
    out.extend(fp.maternal_cousins.iter());
    out.extend(fp.paternal_cousins.iter());
    out
}

/// Build the grader context.
pub fn build_context(d: &AssessmentData) -> GraderContext {
    let manchester_score = calculate_manchester_score(d);
    let bethesda_met = count_bethesda_met(d);
    let premm5_score = d.targeted_risk_scoring.premm5.external_premm5_percent;
    let tyrer_cuzick_lifetime = d
        .targeted_risk_scoring
        .tyrer_cuzick
        .external_lifetime_risk
        .unwrap_or(0.0);

    let mut affected_first_degree = 0;
    let mut early_onset_under50 = 0;
    let mut paediatric_cancers = 0;
    let mut has_male_breast = false;
    let mut has_ovarian = false;
    let mut has_pancreatic = false;

    let fp = &d.family_pedigree;
    let mut first_degree: Vec<&Relative> = vec![&fp.mother, &fp.father];
    first_degree.extend(fp.siblings.iter());
    first_degree.extend(fp.children.iter());

    for r in &first_degree {
        if r.affected_with_cancer == "yes" && !r.cancers.is_empty() {
            affected_first_degree += 1;
        }
    }

    let all = flatten_pedigree(d);
    for r in &all {
        for c in &r.cancers {
            let t = c.kind.to_lowercase();
            if let Some(age) = c.age_at_diagnosis {
                if age <= 50 {
                    early_onset_under50 += 1;
                }
                if age < 18 {
                    paediatric_cancers += 1;
                }
            }
            if t.contains("male breast") || (t.contains("breast") && r.sex == "male") {
                has_male_breast = true;
            }
            if t.contains("ovar") {
                has_ovarian = true;
            }
            if t.contains("pancrea") {
                has_pancreatic = true;
            }
        }
    }

    let mut has_bilateral_breast = false;
    for c in &d.personal_medical_history.cancers {
        let t = c.kind.to_lowercase();
        if t.contains("breast") && c.bilateral == "yes" {
            has_bilateral_breast = true;
        }
        if let Some(age) = c.age_at_diagnosis {
            if age <= 50 {
                early_onset_under50 += 1;
            }
            if age < 18 {
                paediatric_cancers += 1;
            }
        }
        if t.contains("ovar") {
            has_ovarian = true;
        }
        if t.contains("pancrea") {
            has_pancreatic = true;
        }
    }
    let has_multiple_primaries =
        d.personal_medical_history.multiple_primary_cancers == "yes";

    GraderContext {
        manchester_score,
        bethesda_met,
        premm5_score,
        tyrer_cuzick_lifetime,
        affected_first_degree,
        early_onset_under50,
        paediatric_cancers,
        has_male_breast,
        has_ovarian,
        has_pancreatic,
        has_bilateral_breast,
        has_multiple_primaries,
    }
}

/// Evaluate every rule against the data, returning the fired rules.
pub fn run_rules(data: &AssessmentData) -> (Vec<FiredRule>, String) {
    let ctx = build_context(data);
    let mut fired: Vec<FiredRule> = Vec::new();
    let mut level = "low".to_string();
    for rule in all_rules() {
        let severity = (rule.evaluate)(data, &ctx);
        if severity == "low" || severity == "moderate" || severity == "high" {
            fired.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                description: rule.description.to_string(),
                severity: severity.clone(),
            });
            level = max_risk_level(&level, &severity);
        }
    }
    (fired, level)
}

/// Pure function: grade a genetics assessment.
pub fn grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let ctx = build_context(data);
    let (fired_rules, risk_level) = run_rules(data);
    let additional_flags = detect_additional_flags(data);

    GradingResult {
        risk_level,
        manchester_score: ctx.manchester_score,
        bethesda_met: ctx.bethesda_met,
        premm5_score: ctx.premm5_score,
        tyrer_cuzick_lifetime: ctx.tyrer_cuzick_lifetime,
        fired_rules,
        additional_flags,
        timestamp,
    }
}

File: forms/genetics-assessment/front-end-with-svelte/src/lib/engine/genetics-grader.ts
// Genetics Assessment grader. Pure functions: take an AssessmentData object,
// compute the Manchester score, count Bethesda criteria met, extract PREMM5 and
// Tyrer-Cuzick external scores, build a grader context, run every rule, and
// resolve the final risk level as the maximum severity of any fired rule.

import type {
	AssessmentData,
	FiredRule,
	GraderContext,
	GradingResult,
	Relative,
	RiskLevel
} from './types';
import { rules } from './rules';
import { detectAdditionalFlags } from './flagged-issues';
import { maxRiskLevel } from './utils';

// Manchester scoring weights (Evans 2004 / Evans & Lalloo). Per-cancer points.
const MAN_WEIGHTS = {
	femaleBreastUnder30: 6,
	femaleBreast30to39: 4,
	femaleBreast40to49: 3,
	maleBreast: 8,
	ovarianUnder60: 8,
	pancreaticUnder60: 1,
	prostateUnder60: 1
};

function nz(n: number | null | undefined): number {
	return n === null || n === undefined || isNaN(n) ? 0 : Number(n);
}

/** Compute the Manchester score for BRCA1/2 from per-cancer counts. */
export function calculateManchesterScore(d: AssessmentData): number {
	const m = d.targetedRiskScoring.manchester;
	let score = 0;
	// Proband contributions
	score += nz(m.probandFemaleBreastUnder30) * MAN_WEIGHTS.femaleBreastUnder30;
	score += nz(m.probandFemaleBreast30to39) * MAN_WEIGHTS.femaleBreast30to39;
	score += nz(m.probandFemaleBreast40to49) * MAN_WEIGHTS.femaleBreast40to49;
	score += nz(m.probandOvarianUnder60) * MAN_WEIGHTS.ovarianUnder60;
	score += nz(m.probandMaleBreast) * MAN_WEIGHTS.maleBreast;
	// Relative contributions
	score += nz(m.relativeFemaleBreastUnder30) * MAN_WEIGHTS.femaleBreastUnder30;
	score += nz(m.relativeFemaleBreast30to39) * MAN_WEIGHTS.femaleBreast30to39;
	score += nz(m.relativeFemaleBreast40to49) * MAN_WEIGHTS.femaleBreast40to49;
	score += nz(m.relativeOvarianUnder60) * MAN_WEIGHTS.ovarianUnder60;
	score += nz(m.relativeMaleBreast) * MAN_WEIGHTS.maleBreast;
	score += nz(m.relativePancreaticUnder60) * MAN_WEIGHTS.pancreaticUnder60;
	score += nz(m.relativeProstateUnder60) * MAN_WEIGHTS.prostateUnder60;
	return score;
}

/** Count how many of the five Revised Bethesda criteria the patient meets. */
export function countBethesdaMet(d: AssessmentData): number {
	const b = d.targetedRiskScoring.bethesda;
	let n = 0;
	if (b.crcUnder50 === 'yes') n++;
	if (b.synchronousMetachronous === 'yes') n++;
	if (b.msiHistology === 'yes') n++;
	if (b.firstDegreeLynchTumour === 'yes') n++;
	if (b.multipleRelativesLynch === 'yes') n++;
	return n;
}

/** Iterate every relative in the pedigree. */
function flattenPedigree(d: AssessmentData): Relative[] {
	const fp = d.familyPedigree;
	const fixed = [
		fp.maternalGrandmother,
		fp.maternalGrandfather,
		fp.paternalGrandmother,
		fp.paternalGrandfather,
		fp.mother,
		fp.father
	];
	return fixed.concat(
		fp.maternalAuntsUncles || [],
		fp.paternalAuntsUncles || [],
		fp.siblings || [],
		fp.children || [],
		fp.maternalCousins || [],
		fp.paternalCousins || []
	);
}

/** Build the context object that rules consume. */
export function buildContext(d: AssessmentData): GraderContext {
	const manchesterScore = calculateManchesterScore(d);
	const bethesdaMet = countBethesdaMet(d);
	const premm5Raw = d.targetedRiskScoring.premm5.externalPREMM5Percent;
	const premm5Score = premm5Raw === null || premm5Raw === undefined ? null : Number(premm5Raw);
	const tcRaw = d.targetedRiskScoring.tyrerCuzick.externalLifetimeRisk;
	const tyrerCuzickLifetime = tcRaw === null || tcRaw === undefined ? 0 : Number(tcRaw);

	const all = flattenPedigree(d);
	let affectedFirstDegree = 0;
	let earlyOnsetUnder50 = 0;
	let paediatricCancers = 0;
	let hasMaleBreast = false;
	let hasOvarian = false;
	let hasPancreatic = false;

	const fp = d.familyPedigree;
	const firstDegree = [fp.mother, fp.father]
		.concat(fp.siblings || [])
		.concat(fp.children || []);

	for (const r of firstDegree) {
		if (r.affectedWithCancer === 'yes' && (r.cancers || []).length > 0) {
			affectedFirstDegree++;
		}
	}

	for (const r of all) {
		for (const c of r.cancers || []) {
			const type = (c.type || '').toLowerCase();
			const age = c.ageAtDiagnosis;
			if (age !== null && age !== undefined && Number.isFinite(Number(age))) {
				if (Number(age) <= 50) earlyOnsetUnder50++;
				if (Number(age) < 18) paediatricCancers++;
			}
			if (type.includes('male breast') || (type.includes('breast') && r.sex === 'male')) {
				hasMaleBreast = true;
			}
			if (type.includes('ovar')) hasOvarian = true;
			if (type.includes('pancrea')) hasPancreatic = true;
		}
	}

	// Proband history
	const probandCancers = d.personalMedicalHistory.cancers || [];
	let hasBilateralBreast = false;
	for (const c of probandCancers) {
		const type = (c.type || '').toLowerCase();
		const age = c.ageAtDiagnosis;
		if (type.includes('breast') && c.bilateral === 'yes') hasBilateralBreast = true;
		if (age !== null && age !== undefined && Number.isFinite(Number(age))) {
			if (Number(age) <= 50) earlyOnsetUnder50++;
			if (Number(age) < 18) paediatricCancers++;
		}
		if (type.includes('ovar')) hasOvarian = true;
		if (type.includes('pancrea')) hasPancreatic = true;
	}
	const hasMultiplePrimaries = d.personalMedicalHistory.multiplePrimaryCancers === 'yes';

	return {
		manchesterScore,
		bethesdaMet,
		premm5Score,
		tyrerCuzickLifetime,
		affectedFirstDegree,
		earlyOnsetUnder50,
		paediatricCancers,
		hasMaleBreast,
		hasOvarian,
		hasPancreatic,
		hasBilateralBreast,
		hasMultiplePrimaries
	};
}

/**
 * Run every rule and return the fired rules, computed scores, additional flags
 * and the final risk level (maximum severity of any fired rule; 'low' if none).
 */
export function gradeGenetics(data: AssessmentData): GradingResult {
	const ctx = buildContext(data);
	const firedRules: FiredRule[] = [];
	let level: RiskLevel = 'low';

	for (const rule of rules) {
		try {
			const severity = rule.evaluate(data, ctx);
			if (severity === 'low' || severity === 'moderate' || severity === 'high') {
				firedRules.push({
					id: rule.id,
					category: rule.category,
					description: rule.description,
					severity
				});
				level = maxRiskLevel(level, severity);
			}
		} catch (e) {
			console.warn(`Genetics rule ${rule.id} evaluation failed:`, e);
		}
	}

	return {
		riskLevel: level,
		manchesterScore: ctx.manchesterScore,
		bethesdaMet: ctx.bethesdaMet,
		premm5Score: ctx.premm5Score,
		tyrerCuzickLifetime: ctx.tyrerCuzickLifetime,
		firedRules,
		additionalFlags: detectAdditionalFlags(data),
		timestamp: new Date().toISOString()
	};
}
